src/Package/Impl/Sql/Publish/SqlPublishOptionsDialogViewModel.cs

```diff
@@ -0,0 +1,315 @@
+        private readonly IWritableSettingsStorage _settingsStorage;
+        private readonly IProjectConfigurationSettingsProvider _pcsp;
+
+        private bool _canPublish;
+        private bool _targetHasName;
+        private bool _generateTable;
+        private int _selectedTargetIndex;
+        private int _selectedTargetTypeIndex;
+        private int _selectedCodePlacementIndex;
+        private int _selectedQuoteTypeIndex;
+        private string _targetTooltip;
+
+        class DbConnectionData {
+            public string Name;
+            public string ConnectionString;
+        }
+
+        private IReadOnlyList<string> _targets = new List<string>();
+        private IReadOnlyList<string> _targetProjects = new List<string>();
+        private IReadOnlyList<DbConnectionData> _targetConnections = new List<DbConnectionData>();
+
+        public Task InitializationTask { get; private set; } = Task.CompletedTask;
+
+        /// <summary>
+        /// Target types: DACPAC, database, project
+        /// </summary>
+        public IReadOnlyList<string> TargetTypeNames { get; private set; }
+        public int SelectedTargetTypeIndex {
+            get { return _selectedTargetTypeIndex; }
+            set {
+                if (_selectedTargetTypeIndex != value) {
+                    _selectedTargetTypeIndex = value;
+                    Settings.TargetType = TargetTypeFromName(TargetTypeNames[_selectedTargetTypeIndex]);
+                    SetProperty(ref _selectedTargetTypeIndex, value);
+                    PopulateTargets();
+                    UpdateState();
+                }
+            }
+        }
+
+        /// <summary>
+        /// Code placement: inline or table
+        /// </summary>
+        public IReadOnlyList<string> CodePlacementNames { get; private set; }
+        public int SelectedCodePlacementIndex {
+            get { return _selectedCodePlacementIndex; }
+            set {
+                if (_selectedCodePlacementIndex != value) {
+                    _selectedCodePlacementIndex = value;
+                    Settings.CodePlacement = (RCodePlacement)value;
+                    SetProperty(ref _selectedCodePlacementIndex, value);
+                    UpdateState();
+                }
+            }
+        }
+
+        /// <summary>
+        /// SQL name quoting type: none, square brackets or double quotes.
+        /// </summary>
+        public IReadOnlyList<string> QuoteTypeNames { get; private set; }
+        public int SelectedQuoteTypeIndex {
+            get { return _selectedQuoteTypeIndex; }
+            set {
+                if (_selectedQuoteTypeIndex != value) {
+                    Settings.QuoteType = (SqlQuoteType)value;
+                    SetProperty(ref _selectedQuoteTypeIndex, value);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Publishing targets
+        /// </summary>
+        public IReadOnlyList<string> Targets {
```

Comment:
>Targets [](start = 37, length = 7)

It is usually a bad idea to replace collection that is used as a data source for `ItemsControl`. Not only it is a slow operation, it also not handled very well, so you may end up with unbound UI.

Please use `BatchObservableCollection` (use `RPackageManagerViewModel` for an example)